athena/etc/track/track.8
.ll 6i
.TH TRACK 8 "18 April 1988"
.UC 4
.SH NAME
track \- maintain files between machines
.SH SYNOPSIS
.B track
[
.B \- FSTcdfimnpqsvwx
]
.I subscriptionlist
.SH DESCRIPTION
.B Track
supports a
.I librarian/subscriber
model of file update.
That is, subscribing machines are able, but not obliged, to stay in synch with
librarian machines' file-system contents.
.PP
The file
.I subscriptionlist
contains a description of those files, directories, devices, & links
which are to be kept up-to-date on the subscriber machines.
The
.I subscriptionlist's
file-system-description must follow a particular syntax,
which is described in subscriptionlist(5).
For a prototype, see the usual default, /srvd/usr/athena/lib/slists/sys_rvd.
The general default is <workdir>/slists/sys_rvd, where <workdir> is
the "local working directory", as described under the -W option below.
.PP
When
.B track
is invoked with the
.B \-w
option (i.e. as the
.I librarian
machine),
.B track
follows the
.I subscriptionlist's
directions to traverse the librarian's file-system, and creates a 
.I statfile(5)
containing currentness information about the files that are available to
subscribing machines.
The
.I statfile's
default path is /srvd/etc/athena/lib/stats/sys_rvd.
The
.I currentness
information recorded in the statfile includes
file-type, mod-time, uid, gid, permissions, device-type, link-value & checksum.
This is a performance-hack, allowing the subscribers to avoid
net-accesses except when librarian files need to be copied.
The
.I statfile
is not particularly legible.
.PP
When
.B track
is invoked without
.B \-w
(i.e. as the
.I subscriber
machine),
the
.I librarian
machine is specified as an NFS mount-point with the -F option
(default: /srvd).
.B Track
uses the
.I subscriptionlist
to decide what local files are to be considered for possible updating.
The
.I statfile
represents the librarian's exported files, and
.B track
uses it to decide whether each local file should be updated:
any differences in currentness will cause
.B track
to update the local file from the librarian's version.
.PP
It is possible for the librarian & subscribers to use different
.I subscriptionlists.
In general, the librarian "exports" a catalog (the 
.I statfile
), and each subscriber may "import" as much or as little of this catalog's
offering as he wishes.
If the librarian's & subscribers'
.I subscriptionlists
differ, only those files which are specified in both lists can be updated.
.PP
.B Track
is unforgiving when writing a statfile, so as to prevent confused updates.
It tries to forgive errors when updating, so as to discourage partial updates.
Any errors are logged in the file /usr/adm/TRACKLOG.
The TRACKLOG is useful for patching an update that breaks.
.PP
The most commonly-used options are:
.TP
.B \-F<fromdir>
Specify the source ( librarian) root-directory or mount-point.
The default value is /srvd.
.TP
.B \-I
Ignore file-protections ( uid, gid, mode-bits) when tracking as a subscriber.
However, if track creates a file on the subscriber,
the copy will get the librarian's copy's protections.
.TP
.B \-T<todir>
Specify the destination root directory, or mount-point. The default is /.
.TP
.B \-W<workdir>
Change the local working directory (i.e. the directory on the local machine) to
.I workdir.
The default is /mnt/etc/athena/lib, where /mnt is the librarian mount-point.
The working directory is the place where
.B track
will look for the
.I subscriptionlist
(default: slists/sys_rvd),
and the
.I statfile
(default: stats/sys_rvd).
In order to write a statfile (-w option),
.B track
must have permission to write in the local working directory.
.TP
.B \-c
When updating,
compare each regular file's checksum with
the corresponding checksum on the librarian-mountpoint.
The librarian's checksums are always recorded in the
.I statfile.
-c slows the update,
but the checksums enable track to detect file-system corruption.
.TP
.B \-f
Force the update.  Don't be gentle about what is destroyed.  Normally
.B track
is very cautious about files being destroyed by an update.
.TP
.B \-n
Do nothing, just produce a list of files that need updating (disabled by \-w),
and the anticipated consequences of doing the update.
This output goes to stderr, but not into the TRACKLOG.
.TP
.B \-s <statfile>
.B \-s -
Use the named file as the statfile, or use stdio if "-s -" is used.
The default is <workdir>/stats/sys_rvd, where <workdir> is the working-dir
specified by the -W option.
.TP
.B \-v
Produces a very verbose description of what is happening during the update:
which files get updated, and why. This output goes to stderr, but not
into the TRACKLOG.
.TP
.B \-w
Write out a
.I statfile.
This flag is always used by the
.I librarian
machine and never used by the
.I subscriber
machine(s).
This option turns on -c.
.PP
The less commonly-used options:
.TP
.B \-d
Include devices and special files in update. 
.TP
.B \-m[recipient]
( Currently broken) Mail stdout and stderr to the recipient.
The default recipient is root.
.TP
.B \-p
Parse only and produce a verbose description of the fields in the
.I subscriptionlist.
This can be useful when
.B track
complains about syntax errors
in the subscriptionlist.
The output goes to stderr.
.TP
.B \-q
Be quiet, i.e. don't complain about non-fatal errors.
The most common non-fatal error occurs when the
.I librarian's
statfile is out-of-date, so that the subscriber tries to import files
which no longer exist on the
.I librarian.
.TP
.B \-x
Display debugging information ( minimal).
.TP
.B \-S <max_path_depth>
Specify deeper path stacks; this is an internal implementation parameter.
.B Track
is supposed to tell you when to use -S, and how to use it.
Since the default depth accomodates pathnames with ~50 components,
this option is seldom used.
.SH EXAMPLES
.in 8
.ti -8
track
.ti -4
Update the local root's contents from the corresponding contents of
/srvd, according to the constraints specified in the default subscriptionlist
& statfile. This is what the Athena script update_ws does.
.ti -8
track \-p daily
.ti -4
Parse the
.I subscriptionlist
named "daily" and print a verbose
listing of the contents. This command is very helpful when trying to debug
freshly modified
.I subscriptionlists. 
.ti -8
track \-w hourly 
.ti -4
Act as the
.I librarian
and output a
.I statfile
for the
.I subscriptionlist
named "hourly".
.sp 1
.ti -8
track hourly \-w \-q
.ti -4
Same as above, but non-fatal errors will be ignored.
.sp 1
.ti -8
attach paris:root
.br
.ti -8
track daily -F/paris/root
.ti -4
Act as a
.I subscriber
machine and get the updates from root directory on the machine named "paris".
.sp 1
.ti -8
track \-n daily
.ti -4
Act as a
.I subscriber,
but do not perform any updates.
Instead, print a listing of the updates that should be made.  This is analogous
to "make \-n".  Invoking 
.B track
with the
.B \-n
flag is a safe way to preview what will happen before
.B track
is turned loose on a new machine.
.sp 1
.ti -8
track -w sys_rvd -s - -F/source \\
.ti -8
.br
 | track sys_rvd -s - -F/source -T/target -c
.ti -4
Backup to /target those of /source's contents that are specified in sys_rvd,
but only if /target has a different version.
It is best to run this on /target's machine, so that only reads are networked.
.in -8
.SH FILES
 /source/4.3/etc.athena/track/*
.br
 /srvd/etc/athena/lib/slists/sys_rvd
.br
 /srvd/etc/athena/lib/stats/sys_rvd
.br
 /usr/adm/TRACKLOG
.br
 /tmp/sys_rvd.started
.SH SEE ALSO
subscriptionlist(5), statfile(5), permissions(5), cron(8)
.PP
When Network File Systems Aren't Enough: Automatic File Distribution Revisited
.br
.ti 5
by Daniel Nachbar
.SH AUTHOR
Daniel Nachbar (AT&T);
rewritten for bug-fixing, speed, -c option, & extra link-support
by Don Davis (Athena); NFS-conversion done by Rob French (Athena).
.SH BUGS
.br
-m option doesn't work.
.sp 1
If, under -n, the subscriptionlist-entry's topmost target directory
doesn't exist,
the simulated update raises many more errors than would the true update.
.sp 1
Needs a -C (comparison mount-point) option.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/etc/track/track.c
/*
 *	$Id: track.c,v 4.26 2001-08-01 13:47:15 ghudson Exp $
 */

#ifndef lint
static char *rcsid_header_h = "$Id: track.c,v 4.26 2001-08-01 13:47:15 ghudson Exp $";
#endif

#include "bellcore-copyright.h"
#include "mit-copyright.h"

#include "track.h"

char admin[WORDLEN] = DEF_ADM;		/* track administrator */
char workdir[LINELEN];			/* working directory under src/dest
					 * root where slists, statfiles, etc.
					 * can be found */
char binarydir[LINELEN] = DEF_BINDIR;	/* directory in working dir to
					 * find executables */
char fromroot[LINELEN] = DEF_FROMROOT;	/* Root directory for source */
char toroot[LINELEN] = DEF_TOROOT;	/* Root directory for destination */

char lockpath[LINELEN];			/* starting lock filename */
char logfilepath[LINELEN] = DEF_LOG;	/* default log file */
FILE *logfile = NULL;			/* the logfile */
char subfilename[LINELEN] = DEF_SUB;	/* default subscription file */
char subfilepath[LINELEN] = "";		/* alternate subscription file */
char statfilepath[LINELEN];		/* pathname to statfile */
FILE *statfile;				/* the statfile! */
char **statfilebufs;			/* array of lines for sort() */
int cur_line;				/* index into statfilebufs	*/
unsigned maxlines = 0;			/* max # lines in statfile buf array.
					 * write_statline maintains maxlines */
unsigned stackmax = STACKMAX;		/* max depth of pathname-stack vars */

char prgname[LINELEN];

extern int errno;		/* global error number location */
char errmsg[LINELEN];

int writeflag = 0;	/* if set, translate subscription list -> statfile,
			 * rather than pulling files */
int parseflag = 0;	/* if set, just parse the subscription list */
int forceflag = 0;	/* if set, will over-ride lock files */
int verboseflag = 0;	/* if set, files listed on stdout as they're updated */
int cksumflag = 0;	/* if set, compare file checksums when updating */
int nopullflag = 0;	/* if set, find out the differences,
			 *	   but don't pull anything */
int quietflag = 0;	/* if set, don't print non-fatal error messages */
int interactive = 1;	/* if set, don't send errors via mail, print them */
int uflag = NO_CLOBBER;	/* if set, copy a older file on top of a newer one */
int debug = 0;		/* if set, print debugging information */
int ignore_prots = 0;	/* if set, don't use or set uid/gid/mode-bits */
int incl_devs = 0;	/* if set, include devices in update */

/* initialize the global entry-counters with bad values,
 * so that printmsg() can detect them.
 */
Entry entries[ ENTRYMAX];	/* Subscription list entries */
int entrycnt = -1;		/* Number of entries */
int entnum = -1;		/* Current entry number */

void cleanup(), readstat(), clearlocks(), writestat(), walk_trees();
void get_arg(), justshow(), setuperr(), build_path(), openstat();
void closestat();

main(argc,argv)
int argc;
char **argv;
{
	char	scratch[LINELEN];
	int	i;
	struct sigaction sa;

	strcpy(prgname,argv[0]);
	strcpy(errmsg,"");

	umask(022);	/* set default umask for daemons */

	sa.sa_flags = 0;
	sigemptyset(&sa.sa_mask);
	sa.sa_handler = cleanup;

	sigaction( SIGINT, &sa, NULL);
	sigaction( SIGHUP, &sa, NULL);
	sigaction( SIGPIPE, &sa, NULL);

	for(i=1;i<argc;i++) {
		if (argv[i][0] != '-') {
			strcpy( subfilepath, argv[i]);
			continue;
		}
		switch (argv[i][1]) {
		/* -F dirname
		 *    Specify source "root" directory.
		 */
		case 'F':
			get_arg(scratch,argv,&i);
			if (*scratch != '/') {
				getcwd( fromroot, sizeof(fromroot));
				strcat( strcat( fromroot, "/"), scratch);
			}
			else if (! scratch[1]) *fromroot = '\0';
			else strcpy( fromroot, scratch);
			break;

		/* -I
		 *    Ignore protections (uid,gid,mode) when tracking,
		 *    except when creating a file: then, use remote prots.
		 */
		case 'I':
			ignore_prots = 1;
			break;

		/* -S stackmax
		 *    Specify deeper path stacks
		 */ 
		case 'S':
			get_arg(scratch,argv,&i);
			sscanf( scratch, "%d", &stackmax);
			break;

		/* -T dirname
		 *    Specify destination "root" directory.
		 */ 
		case 'T':
			get_arg(scratch,argv,&i);
			if (*scratch != '/') {
				getcwd( toroot, sizeof(toroot));
				strcat( strcat( toroot, "/"), scratch);
			}
			else if (! scratch[1]) *toroot = '\0';
			else strcpy( toroot, scratch);
			break;

		/* -W dirname
		 *    Specify the working directory for
		 *    accessing the subscription-list and statfile.
		 */
		case 'W':
			get_arg(workdir,argv,&i);
			break;
		/* -c
		 *    compare checksums of regular files, when updating.
		 *    the checksums are used to detect file-system
		 *    corruption.
		 */
		case 'c':
			cksumflag = 1;
			break;
		/* -d
		 *    Include devices in an update.
		 */
		case 'd':
			incl_devs = 1;
			break;
		/* -f
		 *    Force updating regardless of locks.
		 */
		case 'f':
			forceflag = 1;
			break;

		/* -m {user}
		 *    Send mail to root/user instead of
		 * displaying messages on the terminal.
		 */
		case 'm':
			interactive = 0;
			get_arg(admin,argv,&i);
			break;
		/* -n
		 *    Produce a list of files that need updating,
		 * but don't actually do anything about them.
		 */
		case 'n':
			nopullflag = 1;
			verboseflag = 1;
			fprintf( stderr, "-n: what we WOULD do:\n");
			break;
		/* -p
		 *    Parse only.  Display a detailed list of the fields in the
		 * subscription file.
		 */
		case 'p':
			parseflag = 1;
			break;
		/* -q
		 *    Be quiet about warning messages.
		 */
		case 'q':
			quietflag = 1;
			break;
		/* -s {pathname}
		 *   use specified file as statfile,
		 * or use stdio if pathname is "-".
		 */
		case 's':
			get_arg( statfilepath, argv, &i);
			break;
		/* -u
		 *    Copy over files regardless of which is newer.
		 */
		case 'u':
			uflag = DO_CLOBBER;
			break;
		/* -v
		 *    Explain what is going on verbosely.
		 */
		case 'v':
			verboseflag = 1;
			break;
		/* -w
		 *    Create a statfile.
		 */
		case 'w':
			writeflag = 1;
			break;
		/* -x
		 *    Display debugging information.
		 */
		case 'x':
			debug = 1;
			break;
		/*
		 * Something isn't right if we got this far...
		 */
		default:
			fprintf(stderr,"track error: bad option %s\n",argv[i]);
			break;
		}
	}

	/*
	 * Set up nullmail interface if not an interactive session.
	 */
	if (!interactive)
		setuperr();

	/* check for existence of root directories:
	 * we shouldn't create them, as they are likely to be remote,
	 * so that the user may have forgotten to attach them.
	 */
	if ( *fromroot && access( fromroot, 0)) {
		sprintf(errmsg,"can't access source root-directory %s\n",
			fromroot);
		do_panic();
	}
	if ( !writeflag && *toroot && access( toroot, 0)) {
		sprintf(errmsg,"can't access target root-directory %s\n",
			toroot);
		do_panic();
	}
	build_path( fromroot, workdir, DEF_SLISTDIR, subfilepath);
	build_path( fromroot, workdir, DEF_STATDIR, statfilepath);

	if ( !quietflag || verboseflag) {
		fprintf( stderr, "using %s as subscription-list\n",
			 subfilepath);
		fprintf( stderr, "using %s as statfile\n",
			 statfilepath);
	}

	/*
	**	redirect yacc/lex i/o
	*/
	parseinit( opensubfile( subfilepath));
	if (yyparse()) {
		strcpy(errmsg,"parse aborted.\n");
		do_panic();
	}
	if (debug)
		printf("parse worked\n");

	sort_entries();

	if (parseflag) {  /* -p: Just show the fields */
		justshow();
		cleanup();
	}

	setlock();

	openstat( statfilepath, writeflag);

	if ( writeflag)		/* -w: Write the exporting statfile */
		writestat();
	else {
		/* update in two passes: links & their parent-dirs first,
		 * which frees up file-system space when links replace files,
		 * then everything else. re-update dirs in second pass,
		 * to facilitate statfile-traversal.
		 */
		readstat( "ld");
		rewind( statfile);
		readstat( "fdbc");
	}
	closestat();

	clearlocks();
	exit( 0);
}			/* end of main() */

#define pathtail( p) p[1+*(int*)p[CNT]]

void
readstat( types) char *types; {
	struct currentness rem_currency, *cmp_currency;
	char statline[ LINELEN], *remname;
	char **from, **to, **cmp;
	char *tail = NULL;

	from = initpath( fromroot);
	to   = initpath(   toroot);
	cmp  = initpath(   toroot);

	/* prime the path-stacks for dec_entry() to
	 * pop the "old" entry-names off.
	 */
	pushpath( from, ""); pushpath( to, ""); pushpath( cmp, "");

	init_next_match();

	while ( NULL != fgets( statline, sizeof statline, statfile)) {

		/* XXX : needs data-hiding work, but will do:
		 *	 only update what main tells us to in this pass.
		 */
		if ( ! strchr( types, *statline)) continue;

		/* extract the currency data from the statline:
		 */
		remname = dec_statfile( statline, &rem_currency);

		/* find the subscription entry corresponding to the
		 * current pathname.
		 * if we reach an entry which is lexicographically greater than
		 * the current pathname, read statfile for the next pathname.
		 * both entries[] & statfile must be sorted by sortkey!
		 */
		if ( 0 >= ( entnum = get_next_match( remname))) continue;

		/* do a breadth-first search of the tree of entries,
		 * to find the entry corresponding to remname:
		 * for example, if /usr & /usr/bin are both entries,
		 * they appear in that order in the entries[] array.
		 * if remname is /usr/bin/foo, we want gettail() to
		 * use /usr/bin's exception-list, not /usr's exception-list.
		 * thus, /usr/bin is the "last match" for /usr/bin/foo.
		 */
		entnum = last_match( remname, entnum);

		tail = remname;
		switch ( gettail( &tail, TYPE( rem_currency.sbuf), entnum)) {
		case NORMALCASE: break;
		case DONT_TRACK: continue;
		case FORCE_LINK: fake_link( fromroot, remname, &rem_currency);
				 break;
		default:	 sprintf(errmsg,"bad value from gettail\n");
				 do_panic();
		}

		/* loosely, tail == remname - fromfile, as
		 * long as tail isn't in the exception-list.
		 * the string remname begins with the string from[ PATH]:
		 * for example: remname =            /usr/bin/foo.
		 *          from[ PATH] =            /usr/bin.
		 *          from[ ROOT] = /mountpoint/usr/bin.
		 * in this example, we get tail == "foo".
		 */

		cmp_currency = dec_entry( entnum, from, to, cmp, tail);

		pushpath( to,  tail);
		pushpath( from, tail);

		if ( ! update_file( cmp_currency, to,
				   &rem_currency, from))
			/* REWRITE:
			do_cmds( entries[entnum].cmdbuf, to[ ROOT])
			 */
			;
		/* remove tail from each path:
		 */
		poppath( to);
		poppath( from);
	}
	/* make sure that the file-systems' superblocks are up-to-date.
	 */
	sync();
	sleep(2);
}

/*
 * Set lock for subscriptionlist file.
 */

setlock()
{
	sprintf( lockpath,"%s/%s.started", DEF_LOCKDIR, subfilename);

	if ( access( lockpath, 0));
	else if ( too_old( lockpath, LOCK_TIME) || forceflag) clearlocks();
	else {
		sprintf( errmsg, "lock set on %s--quitting\n", lockpath);
		do_gripe();
		exit(0);
	}
	if ( close( creat( lockpath,220))) {
		sprintf( errmsg, "can't create lockfile %s\n", lockpath);
		do_panic();
	}
	return(0);
}

/*
 * Erase those locks...
 */

void
clearlocks()
{
	if ( !*lockpath) return;
	if ( unlink( lockpath)) {
		fprintf( stderr, "can't remove lockfile %s", lockpath);
		perror( "system error is: ");
	}
	else if ( verboseflag)
		fprintf( stderr, "cleared lock %s\n",lockpath);
}

/* the array from[] is a set of pointers into a single pathname.
 * it allows us to pass a parsed pathname along in walk_trees()'
 * recursive descent of a directory.
 * the ROOT component is the entire absolute pathname,
 * including the mount-point, for use in file-system calls.
 * the NAME component is the portable pathname, without the mount-point,
 * as the file is described in the statfile.
 * the TAIL component is everything that's added during the recursive descent,
 * for comparison with the exception-list. TAIL lacks both mount-point &
 * the fromfile name.
 */

/*
 * Act like a librarian and write out the statfile.
 */

void
writestat()
{
	char **from, **cmp, **dummy = (char **) NULL;
	struct currentness *entry_currency;

	from = initpath( fromroot);
	cmp  = initpath( fromroot);

	/* prime the path-stacks for dec_entry() to
	 * pop the "old" entry-names off.
	 */
	pushpath( from, ""); pushpath( cmp, "");

	for( entnum = 1; entnum < entrycnt; entnum++) {

		/* dec_entry pushes pathname-qualification
		 * onto the paths 'from' & 'cmp',
		 * and pops when appropriate.
		 */
		entry_currency = dec_entry( entnum, from, dummy, cmp, NULL);

		if (entries[entnum].islink) {
			fake_link( "", from[ NAME], entry_currency);
			write_statline( from, entry_currency);
			continue;
		}

		/* write_statline returns fromfile's true type,
		 * regardless of cmpfile's type:
		 */
		if      ( S_IFDIR != write_statline( from, entry_currency));
		else if ( S_IFDIR != TYPE( entry_currency->sbuf))

			walk_trees( from, dummy, entry_currency);
		else    walk_trees( from, cmp,   entry_currency);

		/* WARNING: walk_trees alters ALL of its arguments */
		/* sort the statfile, and write it out
		 * to the correct directory:
		 */
	}
	sort_stat();
}

/* if the current entry's fromfile is a directory, but its cmpfile isn't,
 * put the same cmpstat in all of fromfile's contents' statlines.
 * if cmpfile is a directory too, its subtree must parallel fromfile's subtree,
 * and each statline reflects the one-to-one (not onto) mapping from
 * fromfile's subtree to cmpfile's subtree:
 * we take fromfile's subnode's pathname ( not including the prefix fromroot),
 * and we take the stat from cmpfile's corresponding subnode.
 */
void
walk_trees( f, c, currency)
char *f[], *c[];
struct currentness *currency;
{
	DIR *dirp;
	struct dirent *dp;
	char *tail;

	dirp = opendir( f[ ROOT]);
	if (!dirp) {
		sprintf(errmsg,"can't open directory %s\n", f[ ROOT]);
		do_gripe();
		return;
	}

	while( dp = readdir( dirp)) {
		if (strcmp(dp->d_name, ".") == 0)
		  continue;
		if (strcmp(dp->d_name, "..") == 0)
		  continue;

		if (! dp->d_ino) continue;    /* empty dir-block */

		pushpath( f, dp->d_name);
		pushpath( c, dp->d_name);

		tail = f[ NAME];
		switch ( gettail( &tail, 0, entnum)) {
		case NORMALCASE: break;
		case FORCE_LINK: fake_link( "", f[ NAME], currency);
				 write_statline( f, currency);
				 /* fall through to poppath() calls */
		case DONT_TRACK: poppath( f);
				 poppath( c);
				 continue;
		default:	 sprintf(errmsg,"bad value from gettail\n");
				 do_panic();
		}
		/* normal case: tail isn't an exception or a forced link.
		 */
		if ( c && get_currentness( c, currency)) {
			sprintf(errmsg,"can't lstat comparison-file %s.\n",
				c[ ROOT]);
			do_panic();
		}
		/* write_statline returns fromfile's type:
		 */
		else if ( S_IFDIR == write_statline( f, currency))
			walk_trees( f, c, currency);

		poppath( f);
		poppath( c);
	}
	closedir(dirp);
}

/*
 * Get a command line argument
 */

void
get_arg(to,list,ptr)
char *to,**list;
int *ptr;
{
	int offset = 2;

	if (strlen(list[*ptr]) == 2) {
		(*ptr)++;
		offset = 0;
	}
	strcpy(to,list[*ptr]+offset);
}

/*
 * Log a message to the logfile.
   UNUSED

log(ptr)
char *ptr;
{
	extern long time();
	extern char *ctime();
	static FILE *logfile = NULL;
	char namebuf[LINELEN];
	char timestring[LINELEN];
	long timebuf;

	if ( NULL == logfile) {
		sprintf( namebuf,"%s/%s",workdir,DEF_LOG);
		if( access( namebuf, 0))
			return;
		if (NULL == (logfile = fopen( namebuf,"a"))) {
			sprintf(errmsg,"can't open log file %s",namebuf);
			do_panic();
		}
	}
	timebuf = time(0);
	strcpy(timestring,ctime(&timebuf));
	timestring[ strlen(timestring)-1] = '\0';
	fprintf(logfile,"%s %s %s\n",timestring,subfilename,ptr);
}
 */

#undef ROOT

/*
 * Execute shell commands
 */

void
do_cmds(cmds,local)
char *cmds,*local;
{
	char *ptr,*nptr;
	FILE *shell;

	if ( ! *cmds) return;
	ptr = cmds;

	shell = popen(DEF_SHELL,"w");
	if (!shell) {
		sprintf(errmsg,"can't open shell %s\n",DEF_SHELL);
		do_gripe();
		return;
	}

	fprintf(shell,"chdir %s\n",toroot);
	fprintf(shell,"%sFILE=%s\n",DEF_SETCMD,local);
	fprintf(shell,"%sROOT=%s\n",DEF_SETCMD,toroot);

	for (;;) {
		nptr = strchr(ptr,'\n');
		if (nptr)
			*nptr = '\0';
		fprintf(shell,"%s\n",ptr);
		if (!nptr) {
			pclose(shell);
			return;
		}
		ptr = nptr+1;
	}
}
/*
 * Show parsing.
 */

void
justshow()
{
	int i,j, size;
	Entry *e;
	List_element *p;

	fprintf( stderr, "subscription-list as parsed:\n\n");

	for (i = 0; i <= entrycnt; i++) {
		e = &entries[ i];
		if ( ! e->fromfile) break;
		fprintf(stderr,
			"entry %d:\n\tislink-- %d\tfromfile-- %s\n",
			i,
			e->islink,
			e->fromfile);
		fprintf(stderr,
			"\tcmpfile-- %s\n\ttofile-- %s\n\tpatterns--\n",
			e->cmpfile,
			e->tofile);
		for( p = e->patterns; p ; p = NEXT( p))
		    fprintf(stderr,"\t\t%s%s\n",
			    FLAG( p) == FORCE_LINK ? "-> " : "",
			    TEXT( p));
		fprintf( stderr, "\texceptions--\n");
		switch( SIGN( e->names.shift)) {
		case -1:
		    for( p = ( List_element *) e->names.table; p ; p = NEXT( p))
			fprintf(stderr,"\t\t%s%s\n",
				FLAG( p) == FORCE_LINK ? "-> " : "",
			 	TEXT( p));
		    fprintf(stderr,
			"track didn't fully parse the exception-list.\n");
		    fprintf(stderr,
			"the most-recently parsed exception was:\n%s%s\n",
			FLAG( e->names.table) == FORCE_LINK ? "-> " : "",
			TEXT( e->names.table));
		    continue;
		case 0: break;
		case 1:
		    size = (unsigned) 0x80000000 >> e->names.shift - 1;
		    for( j = 0; j < size; j++)
		    {
			if ( ! e->names.table[j]) continue;
			fprintf( stderr,"\t\t");
			for ( p = e->names.table[j]; p; p = NEXT( p))
			    fprintf(stderr,"%s%s, ",
				    FLAG( p) == FORCE_LINK ? "-> " : "",
				    TEXT( p));
			fprintf( stderr,"\n");
		    }
		    break;
		}
		fprintf(stderr,"\tcommand-- %s\n",e->cmdbuf);
	}
}

/*
 *	redirect standard error output to mail to the administrator
 *	use nullmail so that null messages won't get sent.
 */

/* REWORK */

void
setuperr()
{
	char msg[LINELEN];
	FILE *tmp;

	sprintf(msg,"%s/nullmail %s",binarydir,admin);
	/*
	**	start a process that will send mail to the adminstrator
	*/
	if ((tmp = popen(msg,"w")) == NULL) {
/*		sprintf( msg, "echo HELP track --%s %s", gargv[0],
			"can't execute nullmail cmd  > /dev/console");
		system(msg); */
		exit(1);
	}
	/*
	**	now connect stderr to the pipe
	*/
	if (dup2(fileno(tmp),2) ==  -1) {
/*		sprintf( msg, "echo HELP track --%s %s", gargv[0],
			"can't dup stderr  > /dev/console");
		system(msg); */
		exit(1);
	}
}

void
build_path( f, w, d, p) char *f, *w, *d, *p; {
	static char buf[ LINELEN];

	if ( ! strcmp( p, "-")) return;
	/*
	 * Get the proper working directory,
	 * where the subscription-list & statfile are.
	 */
	if ( *w) f = "";  /* don't add root-qualification to user's workdir */
	else if ( *p) f = "."; /* don't use default workdir with user's filen */
	else w = DEF_WORKDIR; /* default workdir, default filename */

	if ( *p) d = "";
	else strcpy( p, subfilename);

	sprintf( buf, "%s%s%s%s/%s", f, w, *d ? "/" : "", d, p);
	strcpy( p, buf);
}

FILE *
opensubfile( path) char *path; {
	FILE *subfile;
        if ( ! ( subfile = fopen( path, "r"))) {
                sprintf( errmsg, "Can't open subscriptionlist %s\n", path);
                do_panic();
        }
	return( subfile);
}

void
openstat( path, write) char *path; int write;
{
	char *mode = write? "w"   : "r";
	FILE *std =  write? stdout : stdin;

	if ( ! strcmp( path, "-"))
		statfile = std;
	else if ( ! ( statfile = fopen( path, mode))) {
		sprintf( errmsg, "can't open statfile %s\n", path);
		do_panic();
	}
}

void
closestat() {
	if ( EOF == fclose( statfile)) {
		sprintf( errmsg, "can't close %s\n", statfilepath);
		do_panic();
	}
}

void
cleanup()
{
	clearlocks();
	exit(0);
}
